_posts/2019-04-05-omakase-typescript.markdown

```diff
@@ -0,0 +1,179 @@
+language constraints:
+
+```java
+class Main {
+  public static void main(String[] args) {
+    String name = "Danger";
+    System.out.println("Hello, " + name);
+  }
+}
+```
+
+> Note: if you find the naming of Java and JavaScript confusing (it is, they are two completely separate
+> programming languages, no link at all.) JavaScript was named that way because Java was looking to be really the
+> next hot language (it did turn out that way for a decade or two, but now JavaScript is usually the first language
+> people have heard of.)
+
+Aside from having a lot more lines, the Java version comes with a lot of words that aren't necessarily about
+telling the computer exactly what to do, e.g. `class Main {`, `public static void main(String[] args) {`, `}` and
+`}` again. It also has semi-colons at the end of some lines. Java is aimed at building different things, and these
+extra bits of code make sense within the constraints of building a Java app.
+
+To get to my main point though, there is one standout line I'd like us to compare:
+
+```
+// JavaScript
+var name = "Danger"
+// Java
+String name = "Danger";
+```
+
+Both of these lines declare variables called `name` which contain the value `"Danger"`.
+
+In JavaScript you use the abbreviation `var` to declare a variable. Meanwhile, in Java you need to say _what kind
+of data_ the variable contains. In this case the variable contains a `String`. (A string is a programming term for
+a collection of characters. They `"look like this"`. This [5m video](https://www.youtube.com/watch?v=czTWbdwbt7E)
+is a good primer if you want to learn more.)
+
+Both of these variables contain a string, but the difference is that in Java the variable can _only_ ever contain a
+_string_, because that's what we said when we created the variable. In JS the variable can change to be _anything_,
+like a number, or a list of dates.
+
+To illustrate:
+
+```js
+// Before in JS
+var name = "Danger"
+// Also OK
+var name = 1
+var name = false
+var name = ["2018-02-03", "2019-01-12"]
+
+// Before in Java
+String name = "Danger";
+// Not OK, the code wouldn't be accepted by Java
+String name = 1;
+String name = false
+String name = new String[]{"2018-02-03", "2019-01-12"};
+```
+
+These trade-offs make sense in the context for which these languages were built back in 1995. JavaScript was
+originally designed to be a small programming language which handled simple interactions on websites. Java on the
+other hand was built specifically to make big apps which could run on any computer. Their needs had different
+scales, so the language required programmers write different types of code.
+
+Java required programmers to be more explicit with the values of their variables because the programs they expected
+people to build were more complex. While JavaScript opted for ease of reading, and flexibility because you would be
+able to see most of the code in one go.
```

Comment:
> be able to see most of the code in one go

This ideas needs a little more to it. Java was a language built for enterprise scale and JavaScript was developed when web interactivity was very nascent; is it okay to just say that JavaScript apps had less code because they initially did less stuff than Java apps?

Reply:
agree, done that 👍